src/checkers/inference/BaseInferenceResult.java

```diff
@@ -0,0 +1,39 @@
+package checkers.inference;
+
+import checkers.inference.model.Constraint;
+
+import javax.lang.model.element.AnnotationMirror;
+import java.util.Collection;
+import java.util.Map;
+
+public class BaseInferenceResult implements InferenceResult {
+
+    protected final Map<Integer, AnnotationMirror> inferredResults;
+    protected final Collection<Constraint> unsatisfiableConstraints;
+
+    public BaseInferenceResult(Map<Integer, AnnotationMirror> inferredResults,
+                               Collection<Constraint> unsatisfiableConstraints) {
+        this.inferredResults = inferredResults;
+        this.unsatisfiableConstraints = unsatisfiableConstraints;
+    }
+
+    @Override
+    public boolean doesVariableExist(int varId) {
+        return inferredResults.containsKey(varId);
+    }
+
+    @Override
+    public AnnotationMirror getAnnotation(int varId) {
+        return inferredResults.get(varId);
+    }
+
+    @Override
+    public boolean hasSolution() {
+        return inferredResults != null;
+    }
+
+    @Override
+    public Collection<Constraint> getUnsatisfiableConstraints() {
```

Comment:
Client injects Collection of unsolvable constraints here.
This class doesn't have any invariant, but the client is responsible for following invariants:
For example,
if `hasSolution()` returns true, client needs to pass non-null set of constraint if a backend support explanation; pass null if it doesn't support it.

Reply:
Thanks for the explanation. Could you add an assertion to document this in code?
i.e.

```java
 public Collection<Constraint> getUnsatisfiableConstraints() {
    assert !hasSolution() : "If has solution, client shouldn't call getUnsatisfiableConstraints() as the result is not guaranteed to be non-null!"

     return unsatisfiableConstraints;
}
```

Reply:
I'll do that.